BUG (?): linalg solve behavior change breaking downstream tests

I bisected SciPy against a rather messy reproducer--it requires using some obscure project we have (https://github.com/lanl/ldrd_neat_ml) that started failing tests with SciPy `1.17.0`, and those tests in turn interact with https://github.com/marcotcr/lime (not well maintained I admit) and `scikit-learn`. So, apologies that this isn't cleaner yet--I asked a trainee to help with the root causing. I'm not yet sure that this is the same as gh-24345 since @lucascolley mentions a different commit there--maybe, maybe not... It looks like the commit I'm worried about is from gh-23071, which is the same PR Lucas refers to, though a different commit. So some chance of overlap, but not guaranteed yet.


```
9e1176f105005cd304898f66d889971fc2cd999a is the first bad commit
commit 9e1176f105005cd304898f66d889971fc2cd999a
Author: Evgeni Burovski <evgeny.burovskiy@gmail.com>
Date:   Tue May 27 16:43:10 2025 +0200

    MAINT: linalg/solve: backwards compat changes to the new solve

 scipy/linalg/_basic.py           | 60 +++++++++++++++++++++++++++++++++++---------------
 scipy/linalg/tests/test_basic.py |  5 +++--
 2 files changed, 45 insertions(+), 20 deletions(-)
```

The traceback, admittedly partially propagating through some projects we don't care about here, is pasted below the fold. I don't think it will be much help yet, but we're getting a bunch of `nan`s propagating through now.

<details>

```
_______________________________ test_build_lime_data[0-expected0] _______________________________
[gw4] darwin -- Python 3.12.3 /Users/treddy/python_venvs/py_312_ldrd_neat_dev/bin/python

important_col = 0, expected = [0, 0, 0]

    @pytest.mark.parametrize("important_col, expected", [
        # should be easy to predict the important
        # feature for the non-zero rows
        (0, [0, 0, 0]),
        (1, [1, 1, 1]),
        (2, [2, 2, 2]),
    ])
    def test_build_lime_data(important_col, expected):
        # very simple dataset to check the most basic
        # properties of the LIME data structure generation
        data = np.zeros(shape=(4, 3), dtype=np.float64)
        data[:-1, important_col] = 1.0
        y = np.asarray([1, 1, 1, 0])
        X = pd.DataFrame(data=data, columns=["1", "2", "3"])
        rf = RandomForestClassifier(random_state=0)
        rf.fit(X, y)
>       lime_actual = lib.build_lime_data(X=X, model=rf)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

neat_ml/tests/test_lib.py:144: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/joblib/memory.py:607: in __call__
    return self._cached_call(args, kwargs, shelving=False)[0]
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/joblib/memory.py:562: in _cached_call
    return self._call(call_id, args, kwargs, shelving)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/joblib/memory.py:832: in _call
    output = self.func(*args, **kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
neat_ml/lib.py:792: in build_lime_data
    exp = explainer_lime.explain_instance(X.to_numpy()[index],
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/lime/lime_tabular.py:452: in explain_instance
    ret_exp.score, ret_exp.local_pred) = self.base.explain_instance_with_data(
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/lime/lime_base.py:183: in explain_instance_with_data
    used_features = self.feature_selection(neighborhood_data,
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/lime/lime_base.py:134: in feature_selection
    return self.feature_selection(data, labels, weights,
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/lime/lime_base.py:76: in feature_selection
    return self.forward_selection(data, labels, weights, num_features)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/lime/lime_base.py:61: in forward_selection
    score = clf.score(data[:, used_features + [feature]],
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/base.py:654: in score
    return r2_score(y, y_pred, sample_weight=sample_weight)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/utils/_param_validation.py:218: in wrapper
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/metrics/_regression.py:1281: in r2_score
    _check_reg_targets_with_floating_dtype(
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/metrics/_regression.py:208: in _check_reg_targets_with_floating_dtype
    y_type, y_true, y_pred, sample_weight, multioutput = _check_reg_targets(
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/metrics/_regression.py:115: in _check_reg_targets
    y_pred = check_array(y_pred, ensure_2d=False, dtype=dtype)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/utils/validation.py:1074: in check_array
    _assert_all_finite(
../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/utils/validation.py:133: in _assert_all_finite
    _assert_all_finite_element_wise(
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

X = array([nan, nan, nan, ..., nan, nan, nan])

    def _assert_all_finite_element_wise(
        X, *, xp, allow_nan, msg_dtype=None, estimator_name=None, input_name=""
    ):
        # Cython implementation doesn't support FP16 or complex numbers
        use_cython = (
            xp is np and X.data.contiguous and X.dtype.type in {np.float32, np.float64}
        )
        if use_cython:
            out = cy_isfinite(X.reshape(-1), allow_nan=allow_nan)
            has_nan_error = False if allow_nan else out == FiniteStatus.has_nan
            has_inf = out == FiniteStatus.has_infinite
        else:
            has_inf = xp.any(xp.isinf(X))
            has_nan_error = False if allow_nan else xp.any(xp.isnan(X))
        if has_inf or has_nan_error:
            if has_nan_error:
                type_err = "NaN"
            else:
                msg_dtype = msg_dtype if msg_dtype is not None else X.dtype
                type_err = f"infinity or a value too large for {msg_dtype!r}"
            padded_input_name = input_name + " " if input_name else ""
            msg_err = f"Input {padded_input_name}contains {type_err}."
            if estimator_name and input_name == "X" and has_nan_error:
                # Improve the error message on how to handle missing values in
                # scikit-learn.
                msg_err += (
                    f"\n{estimator_name} does not accept missing values"
                    " encoded as NaN natively. For supervised learning, you might want"
                    " to consider sklearn.ensemble.HistGradientBoostingClassifier and"
                    " Regressor which accept missing values encoded as NaNs natively."
                    " Alternatively, it is possible to preprocess the data, for"
                    " instance by using an imputer transformer in a pipeline or drop"
                    " samples with missing values. See"
                    " https://scikit-learn.org/stable/modules/impute.html"
                    " You can find a list of all estimators that handle NaN values"
                    " at the following page:"
                    " https://scikit-learn.org/stable/modules/impute.html"
                    "#estimators-that-handle-nan-values"
                )
>           raise ValueError(msg_err)
E           ValueError: Input contains NaN.

../../../python_venvs/py_312_ldrd_neat_dev/lib/python3.12/site-packages/sklearn/utils/validation.py:182: ValueError
```

</details>

Need to clean up the reproducer...

--- a/scipy/linalg/_basic.py
+++ b/scipy/linalg/_basic.py
@@ -277,6 +277,9 @@ def solve(a, b, lower=False, overwrite_a=False,
         return x
 
     if a_is_scalar:
+        if a1.item() == 0:
+            raise LinAlgError("A singular matrix detected.")
+
         out = b1 / a1
         return out[..., 0] if b_is_1D else out
 

--- a/scipy/linalg/tests/test_basic.py
+++ b/scipy/linalg/tests/test_basic.py
@@ -808,6 +808,22 @@ class TestSolve:
         with (pytest.raises(LinAlgError, match="singular"), np.errstate(all='ignore')):
             solve(A, b, assume_a=structure)
 
+
+    @pytest.mark.parametrize('b', [0, 1, [0, 1]])
+    def test_singular_scalar(self, b):
+        # regression test for gh-24355: scalar a=0 is singular
+        # thus should raise the same error 
+
+        with pytest.raises(LinAlgError):
+            a = np.zeros((1, 1))
+            solve(a, b)
+
+        with pytest.raises(LinAlgError):
+            solve(0, b)
+
+        with pytest.raises(LinAlgError):
+            solve([[0]], b)
+
     def test_multiple_rhs(self):
         a = np.eye(2)
         rng = np.random.default_rng(1234)

